firewood/src/merkle/stream.rs

```diff
@@ -7,399 +7,432 @@ use crate::{
+                // should be visited since they are after `key`.
+                iter_stack.push(IterationNode::Visited {
+                    key: matched_key_nibbles.iter().copied().collect(),
+                    children_iter: Box::new(
+                        get_children_iter(branch)
+                            .filter(move |(_, pos)| *pos > next_unmatched_key_nibble),
+                    ),
+                });
+
+                // Figure out if the child at `next_unmatched_key_nibble` is a prefix of `key`.
+                // (i.e. if we should run this loop body again)
+                #[allow(clippy::indexing_slicing)]
+                let Some(child_addr) = branch.children[next_unmatched_key_nibble as usize] else {
+                    // There is no child at `next_unmatched_key_nibble`.
+                    // We'll visit `node`'s first child at index > `next_unmatched_key_nibble`
+                    // first (if it exists).
+                    return Ok(NodeStreamState::Iterating { iter_stack });
                 };
 
-                self.poll_next(_cx)
-            }
+                matched_key_nibbles.push(next_unmatched_key_nibble);
 
-            IteratorState::Iterating {
-                check_child_nibble,
-                visited_node_path,
-            } => {
-                let next = find_next_result(merkle, visited_node_path, check_child_nibble)
-                    .map_err(|e| api::Error::InternalError(Box::new(e)))
-                    .transpose();
+                let child = merkle.get_node(child_addr)?;
 
-                Poll::Ready(next)
+                let partial_key = match child.inner() {
+                    NodeType::Branch(branch) => &branch.path,
+                    NodeType::Leaf(leaf) => &leaf.path,
+                    NodeType::Extension(_) => {
+                        unreachable!("extension nodes shouldn't exist")
+                    }
+                };
+
+                let (comparison, new_unmatched_key_nibbles) =
+                    compare_partial_path(partial_key.iter(), unmatched_key_nibbles);
+                unmatched_key_nibbles = new_unmatched_key_nibbles;
+
+                match comparison {
+                    Ordering::Less => {
+                        // `child` is before `key`.
+                        return Ok(NodeStreamState::Iterating { iter_stack });
+                    }
+                    Ordering::Equal => {
+                        // `child` is a prefix of `key`.
+                        matched_key_nibbles.extend(partial_key.iter().copied());
+                        node = child;
+                    }
+                    Ordering::Greater => {
+                        // `child` is after `key`.
+                        let key = matched_key_nibbles
+                            .iter()
+                            .copied()
+                            .chain(partial_key.iter().copied())
+                            .collect();
```

Comment:
nit; doubtful it makes any difference but seems easier to digest
```suggestion
                        let key = matched_key_nibbles
                            .iter()
                            .chain(partial_key.iter())
                            .copied()
                            .collect();
```